onnxruntime/core/framework/session_state.h

```diff
@@ -59,6 +61,34 @@ struct MemoryPatternGroup;
 class MemoryInfo;
 #endif
 
+#ifdef ENABLE_TRAINING
+struct PartialGraphExecutionState {
+ public:
+  PartialGraphExecutionState() {
+    execution_frame_ = nullptr;
+  }
+
+  ~PartialGraphExecutionState() = default;
+
+  void SetProgramCounterStart(size_t start) { program_counter_start_ = start; }
+  void SetProgramCounterEnd(size_t end) { program_counter_end_ = end; }
+
+  size_t GetProgramCounterStart() { return program_counter_start_; }
+  size_t GetProgramCounterEnd() { return program_counter_end_; }
+
+  void SetExecutionFrame(std::unique_ptr<ExecutionFrame>& frame) {
+    execution_frame_ = std::move(frame);
```

Comment:
>std::move [](http://example.com/codeflow?start=23&length=9)

This does not seem to be right. std::move should come from the caller, so caller would explicitly release the ownership.
Or avoid std::move, but assign the raw pointer:
execution_frame.reset(frame.release());

Reply:
I believe the correct pattern is to not pass the unique_ptr by reference - pass it by value (caller has to std::move so intent is explicit at the call site). Using 'release' should be avoided.

Reply:
Thanks, Ke and Scott, nice learning for me. Will keep this in mind moving forward. Made the change per Scott's suggestion.